plugins/GSdx/GSRendererDX.cpp

```diff
@@ -147,6 +151,189 @@ void GSRendererDX::EmulateZbuffer()
+		int  pos = (v[0].XYZ.X - o.OFX) & 0xFF;
+		bool write_ba = (pos > 112 && pos < 136);
+		// Read texture is 8 to 16 pixels (same as above)
+		float tw = (float)(1u << m_context->TEX0.TW);
+		int tex_pos = (PRIM->FST) ? v[0].U : (int)(tw * v[0].ST.S);
+		tex_pos &= 0xFF;
+		m_ps_sel.read_ba = (tex_pos > 112 && tex_pos < 144);
+
+		// Convert the vertex info to a 32 bits color format equivalent
+		if (PRIM->FST) {
+
+			for(size_t i = 0; i < count; i += 2) {
+				if (write_ba)
+					v[i].XYZ.X   -= 128u;
+				else
+					v[i+1].XYZ.X += 128u;
+
+				if (m_ps_sel.read_ba)
+					v[i].U       -= 128u;
+				else
+					v[i+1].U     += 128u;
+
+				// Height is too big (2x).
+				int tex_offset = v[i].V & 0xF;
+				GSVector4i offset(o.OFY, tex_offset, o.OFY, tex_offset);
+
+				GSVector4i tmp(v[i].XYZ.Y, v[i].V, v[i+1].XYZ.Y, v[i+1].V);
+				tmp = GSVector4i(tmp - offset).srl32(1) + offset;
+
+				v[i].XYZ.Y   = (uint16)tmp.x;
+				v[i].V       = (uint16)tmp.y;
+				v[i+1].XYZ.Y = (uint16)tmp.z;
+				v[i+1].V     = (uint16)tmp.w;
+			}
+		} else {
+			const float offset_8pix = 8.0f / tw;
+
+			for(size_t i = 0; i < count; i += 2) {
+				if (write_ba)
+					v[i].XYZ.X   -= 128u;
+				else
+					v[i+1].XYZ.X += 128u;
+
+				if (m_ps_sel.read_ba)
+					v[i].ST.S    -= offset_8pix;
+				else
+					v[i+1].ST.S  += offset_8pix;
+
+				// Height is too big (2x).
+				GSVector4i offset(o.OFY, o.OFY);
+
+				GSVector4i tmp(v[i].XYZ.Y, v[i+1].XYZ.Y);
+				tmp = GSVector4i(tmp - offset).srl32(1) + offset;
+
+				//fprintf(stderr, "Before %d, After %d\n", v[i+1].XYZ.Y, tmp.y);
+				v[i].XYZ.Y   = (uint16)tmp.x;
+				v[i].ST.T   /= 2.0f;
+				v[i+1].XYZ.Y = (uint16)tmp.y;
+				v[i+1].ST.T /= 2.0f;
+			}
+		}
+
+		// Please bang my head against the wall!
+		// 1/ Reduce the frame mask to a 16 bit format
+		// FIXME GSVector will be nice here
+		const uint32& m = m_context->FRAME.FBMSK;
+		uint32 fbmask = ((m >> 3) & 0x1F) | ((m >> 6) & 0x3E0) | ((m >> 9) & 0x7C00) | ((m >> 16) & 0x8000);
+		om_bsel.wrgba = 0;
+
+		// 2 Select the new mask (Please someone put SSE here)
+		if ((fbmask & 0xFF) == 0) {
+			if (write_ba)
+				om_bsel.wb = 1;
+			else
+				om_bsel.wr = 1;
+		} else if ((fbmask & 0xFF) != 0xFF) {
+#ifdef _DEBUG
+			fprintf(stderr, "Please fix me! wb %u wr %u\n", om_bsel.wb, om_bsel.wr);
+#endif
+			//ASSERT(0);
+		}
+
+		fbmask >>= 8;
+		if ((fbmask & 0xFF) == 0) {
+			if (write_ba)
+				om_bsel.wa = 1;
+			else
+				om_bsel.wg = 1;
+		} else if ((fbmask & 0xFF) != 0xFF) {
+#ifdef _DEBUG
+			fprintf(stderr, "Please fix me! wa %u wg %u\n", om_bsel.wa, om_bsel.wg);
+#endif
+			//ASSERT(0);
+		}
+
+	} else {
+		//m_ps_sel.fmt = GSLocalMemory::m_psm[m_context->FRAME.PSM].fmt;
+
+		om_bsel.wrgba = ~GSVector4i::load((int)m_context->FRAME.FBMSK).eq8(GSVector4i::xffffffff()).mask();
+	}
+}
+
+void GSRendererDX::EmulateChannelShuffle(GSTexture** rt, const GSTextureCache::Source* tex)
+{
+	dev = (GSDeviceDX*)m_dev;
```

Comment:
This variable doesn't seem to be used in this function or did I miss it.

Reply:
No, it's only used for setting shader resource and cb. Thought it might come in handy in the future when(if) the shaders get ported. I can remove it.

Reply:
Note that I did those cast in OpenGL because I only have a single device. This way the compiler can replace the dynamic selection of the device by a static call.

Here, it might not help the compiler because it would need to chose Dx9 device or Dx11 device.